### ISSUE TITLE
Typescript event.detail

### ISSUE THREAD

**Issue by jdspugh** [framework7io/framework7#2792]
* Framework7 version: 3.4.2
* Platform and Target: All

### Describe the bug
Typescript typing for event.detail is missing.

### To Reproduce
Add some code that uses event.detail

Dom7(document).on("page:init", '.page[data-name="exercise-edit"]', function(
  event: Event
) {
  let page = event.detail;

### Expected behavior
No error.

### Actual Behavior
Type error.

### Solution
Add these lines to lib.dom.d.ts

import{ Dom7 } from "node_modules/framework7/framework7";
import { View } from "node_modules/framework7/components/view/view";
interface PageData {
  app?: object;
  view?: View.View;
  router?: object;
  name?: string;
  el: HTMLElement;
  $el: Dom7.Dom7;
  navbarEl: HTMLElement;
  $navbarEl: Dom7.Dom7;
  from?: string;
  to?: string;
  position: string;
  direction: string;
  route: {
    url: string;
    path: string;
    query: object;
    params: object;
    name: string;
    hash: object;
    route: object;
    context: object;
  };
  pageFrom: object;
  context?: object;
}

interface Event {
    readonly detail: PageData;

**Comment by jdspugh** [comment]
These changes also need to be made in lib.dom.d.ts:

interface CustomEvent extends Event {
    /**
     * Returns any custom data event was created with.
     * Typically used for synthetic events.
     */
    readonly detail: PageData;
    initCustomEvent(typeArg: string, canBubbleArg: boolean, cancelableArg: boolean, detailArg: PageData): void;
}

declare var CustomEvent: {
    prototype: CustomEvent;
    new<T>(typeArg: string, eventInitDict?: CustomEventInit<T>): CustomEvent;
};

interface UIEvent extends Event {
    // readonly detail: number;
    readonly view: Window;
    initUIEvent(typeArg: string, canBubbleArg: boolean, cancelableArg: boolean, viewArg: Window, detailArg: number): void;
}

**Comment by nolimits4web** [comment]
@JasonKleban do you have idea how can we add it?

**Comment by jdspugh** [comment]
Seems the PageData type is also missing and needs to be defined. So I'm using this until a deeper solution is found:

Dom7(document).on("page:init", '.page[data-name="exercise-edit"]', function(
  event: Event
) {
  let page = (<Event & {detail:PageData}>event).detail;
  ...

import { View } from "framework7/components/view/view";
import { Dom7 } from "dom7";
interface PageData {
  app?: object;
  view?: View.View;
  router?: object;
  name?: string;
  el: HTMLElement;
  $el: Dom7;
  navbarEl: HTMLElement;
  $navbarEl: Dom7;
  from?: string;
  to?: string;
  position: string;
  direction: string;
  route: {
    url: string;
    path: string;
    query: object;
    params: object;
    name: string;
    hash: object;
    route: object;
    context: object;
  }
  pageFrom: object;
  context?: object;
}

**Comment by jdspugh** [comment]
Ok I found another solution for this. Using the "declare global" will allow us to extract the "detail" parameter from all Events. But is this what we want? Cause not all events will have the "detail" parameter available.

declare global {
  interface Event {
    detail: PageData;
  }
}
import { View } from "framework7/components/view/view";
import { Dom7 } from "dom7";
interface PageData {
  app?: object;
  view?: View.View;
  router?: object;
  name?: string;
  el: HTMLElement;
  $el: Dom7;
  navbarEl: HTMLElement;
  $navbarEl: Dom7;
  from?: string;
  to?: string;
  position: string;
  direction: string;
  route: {
    url: string;
    path: string;
    query: object;
    params: object;
    name: string;
    hash: object;
    route: object;
    context: object;
  };
  pageFrom: object;
  context?: object;
}

**Comment by nolimits4web** [comment]
Hey @jdspugh. Page Data definition is available in Router module: https://github.com/framework7io/framework7/blob/master/src/core/modules/router/router.d.ts#L140
```
import { Router } from 'framework7/modules/router/router';

// Router.Page -> is what you are looking for

**Comment by jdspugh** [comment]
@nolimits4web Great, thanks. So this code does the trick to be able to access event.detail without any type errors:

import { Router } from 'node_modules/framework7/modules/router/router';
declare global {
  interface Event {
    detail: Router.Page;
  }
}

**Comment by nolimits4web** [comment]
Yeah, but the issue is that `event.detail` not always contains Page Data, like in case Sortable events. So still looking for a solution how to handle this correctly

**Comment by JasonKleban** [comment]
Sorry I haven't had time to dig into this.... would it work if overloads for event methods knows that handlers for page events get Page Data and Sortable events get sortable-related data?  

```ts
  on(eventName : 'page:init', handler : (event : { detail: ... }) => void) : void
...
  on(eventName : 'textarea:resize', handler : ...) : void
  on(event : string, handler : () => void) : void
```

Know what I mean?

**Comment by jdspugh** [comment]
That's the way I imagined it would be done.

**Comment by nolimits4web** [comment]
@JasonKleban yeah, but how can we specify it? I mean, for example, we have `DomEvents` in `router.d.ts` and others but we don't use them anywhere. I guess we need somehow make them extend `Event` type?

**Comment by nolimits4web** [comment]
This issue was incorrectly closed on 2026-05-19 by a compromised credential (Shai-Hulud npm supply-chain worm). Reopening — apologies for the noise.
